python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,237 @@
+        a Ray cluster.
+
+        Args:
+            max_workers:
+                The maximum number of Ray workers to distribute the
+                tasks over. This is achieved by creating a fixed pool
+                of Actors (see
+                https://docs.ray.io/en/latest/ray-core/actors/
+                actor-utils.html#actor-pool).
+
+            All additional keyword arguments are passed to ray.init()
+            (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+            For example, this will connect to a cluster at the specified address:
+            .. code-block:: python
+
+                RayExecutor(address='192.168.0.123:25001')
+
+            Note: excluding an address will initialise a local Ray cluster.
+        """
+        self._shutdown_lock = False
+        self._futures = []
+        self.__actor_pool = None
+        self.__remote_fn = None
+        self.context = None
+
+        """
+        The following is necessary because `@ray.remote` is only available at runtime.
+        """
+
+        if max_workers is None:
+
+            @ray.remote
+            def remote_fn(fn):
+                return fn()
+
+            self.__remote_fn = remote_fn
+        else:
+            if max_workers < 1:
+                raise ValueError("max_workers must be >= 1")
+
+            @ray.remote
+            class ExecutorActor:
+                def __init__(self):
+                    pass
+
+                def actor_function(self, fn):
+                    return fn()
+
+            actors = [
+                ExecutorActor.options(name=f"actor-{i}").remote()
+                for i in range(max_workers)
+            ]
+            self.__actor_pool = ray.util.ActorPool(actors)
+
+        self.context = ray.init(ignore_reinit_error=True, **kwargs)
+
+    def submit(self, fn, *args, **kwargs):
+        """Submits a callable to be executed with the given arguments.
+
+        Schedules the callable to be executed as `fn(*args, **kwargs)` and returns
+        a Future instance representing the execution of the callable.
+
+        Returns:
+            A Future representing the given call.
+
+        Usage example:
+
+        .. code-block:: python
+
+            with RayExecutor() as ex:
+                future_0 = ex.submit(lambda x: x * x, 100)
+                future_1 = ex.submit(lambda x: x + x, 100)
+                result_0 = future_0.result()
+                result_1 = future_1.result()
+        """
+        self._check_shutdown_lock()
+        fn_curried = partial(fn, *args, **kwargs)
+        if self.__actor_pool:
+            self.__actor_pool.submit(
+                lambda a, _: a.actor_function.remote(fn_curried), None
+            )
+            oref = self.__actor_pool._index_to_future[
+                self.__actor_pool._next_task_index - 1
+            ]
+        else:
+            oref = self.__remote_fn.options(name=fn.__name__).remote(fn_curried)
+        future = oref.future()
```

Comment:
Q. I have never used the .future() API from Ray. Will the object reference that is associated with this future GC'ed when the future is GC'ed? Can you verify and lmk?

Reply:
We may need to add callback into the future and clean up `self._futures` when the future is GC'ed? Otherwise the object may never be GC'ed from the object store. E.g., 

```python3
future_0 = ex.submit(lambda x: <return a big object that's bigger than 100KB>)
future.result()
del future
```
After this, the object is not cleaned up from the object store because it is never GC'ed from `self._futures`?

Reply:
I suspect that the addition of `del self._futures` above solves this?

Reply:
Let me try verifying this locally!

Reply:
It appears to clean up futures internally. If the user has a reference to a future in their code, however, the ObjectRef is retained.